document/js-api/index.bs

```diff
@@ -1120,26 +1144,155 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
 1. If |type| is [=f64=],
     1. Let |f64| be ? [=ToNumber=](|v|).
     1. Return [=f64.const=] |f64|.
-1. If |type| is [=funcref=],
+1. If |type| is of the form [=ref=] |null| |heaptype| and is a subtype of [=ref=] |null| [=heap-type/func=],
     1. If |v| is null,
-        1. Return [=ref.null=] [=funcref=].
+        1. If |null| is present,
+            1. Return [=ref.null=] |heaptype|.
+        1. Otherwise,
+            1. Throw a {{TypeError}}.
     1. If |v| is an [=Exported Function=],
         1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
-        1. Return [=ref.func=] |funcaddr|.
+        1. Let |casted| be [=ref_cast=](|heaptype|, |funcaddr|).
+        1. Return |casted|.
     1. Throw a {{TypeError}}.
-1. If |type| is [=externref=],
+1. If |type| is of the form [=ref=] |null| [=extern=],
     1. If |v| is null,
-        1. Return [=ref.null=] [=externref=].
+        1. If |null| is present,
+            1. Return [=ref.null=] [=externref=].
+        1. Otherwise,
+            1. Throw a {{TypeError}}.
     1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
     1. If a [=extern address=] |externaddr| exists such that |map|[|externaddr|] is the same as |v|,
         1. Return [=ref.extern=] |externaddr|.
     1. Let [=extern address=] |externaddr| be the smallest address such that |map|[|externaddr|] [=map/exists=] is false.
     1. [=map/Set=] |map|[|externaddr|] to |v|.
     1. Return [=ref.extern=] |externaddr|.
+1. If |type| is of the form [=ref=] |null| |heaptype| and is a subtype of [=ref=] |null| [=any=],
+    1. If |v| is null,
+        1. If |null| is present,
+            1. Return [=ref.null=] |heaptype|.
+        1. Otherwise,
+            1. Throw a {{TypeError}}.
+    1. If |heaptype| is [=i31=],
```

Comment:
This condition seems wrong: i31's need to be generated for type `any` or `eq` as well, and the conversion should be the same (and accept the same values) regardless of the type. Likewise, you'll need to allow arbitrary JS values for `any`, not just ints or GC objects.

All non-null values ought to be converted uniformly by first performing internalize and then a type check. The internalize function will then have to do the job to convert numbers to i31 if they are within range, and treat them as host values otherwise.

This indeed implies that the JS spec needs to amend the core spec by customising the meaning of internalize/externalize. The core spec cannot really define it, because it doesn't know what host values are. We'll need some hook in the core spec for this host-specific semantics.

Reply:
> I think all non-null values ought to be converted uniformly by first performing internalize and then a type check. 

I've changed it to use to use this approach now: https://github.com/WebAssembly/gc/commit/2d39b2254b8198b49b8e4a03a3de6fcfca18bfe0#diff-73846d0b7899a94fe050d3c0ba447da607bebedf80f8b5ed31cb4d164756f048R1170

Also added a new section in the spec that details the requirements on externalize & internalize: https://github.com/WebAssembly/gc/pull/352/commits/2d39b2254b8198b49b8e4a03a3de6fcfca18bfe0#diff-73846d0b7899a94fe050d3c0ba447da607bebedf80f8b5ed31cb4d164756f048R1348